src/server/player_sao.cpp

```diff
@@ -473,26 +473,31 @@ void PlayerSAO::setHP(s32 hp, const PlayerHPChangeReason &reason, bool send)
 
 	{
 		s32 hp_change = m_env->getScriptIface()->on_player_hpchange(this, hp - m_hp, reason);
-		if (hp_change == 0)
+		if (hp_change == 0) {
+			if (from_client) // Correct client side prediction
+				m_env->getGameDef()->SendPlayerHPOrDie(this, reason);
 			return;
+		}
 
 		hp = m_hp + hp_change;
 	}
 
 	s32 oldhp = m_hp;
 	hp = rangelim(hp, 0, m_prop.hp_max);
 
-	if (hp < oldhp && isImmortal())
-		return; // Do not allow immortal players to be damaged
+	if (hp < oldhp && isImmortal()) { // Do not allow immortal players to be damaged
+		if (from_client) // Correct client side prediction
+			m_env->getGameDef()->SendPlayerHPOrDie(this, reason);
+		return;
+	}
 
 	m_hp = hp;
 
 	// Update properties on death
 	if ((hp == 0) != (oldhp == 0))
 		m_properties_sent = false;
 
-	if (send)
-		m_env->getGameDef()->SendPlayerHPOrDie(this, reason);
+	m_env->getGameDef()->SendPlayerHPOrDie(this, reason);
 }
```

Comment:
I think it'd work out better if you rewrote this function so that it does not return early.

Then at the end you can have central handling that looks like this:
```c
if (from_client) {
    if (client_expects_hp != actual_new_hp) // Correct client side prediction
        SendPlayerHPOrDie();
} else {
    if (old_hp != actual_new_hp)
        SendPlayerHPOrDie();
}
```
This would be much clearer and can be confirmed bug-free from a single look.

Reply:
> `if (client_expects_hp != actual_new_hp)`

If client_expects_hp == 0 and actual_new_hp == 0, you still should send death screen.